File: a2/java/src/Terrain.java
import javax.imageio.ImageIO;
import java.awt.Point;
import java.awt.image.BufferedImage;
import java.io.File;
import java.util.ArrayList;
import java.util.List;
import static java.lang.Math.abs;

public class Terrain implements Searchable, Comparable<Searchable> {

    protected BufferedImage img;
    protected Point current;
    protected Point goal;
    protected double cost;
    private int heur;

    public Terrain(BufferedImage img, Point current, Point goal) {
        this.img = img;
        this.current = current;
        this.goal = goal;
    }

    @Override
    public boolean isDone() {
        return current.equals(goal);
    }

    @Override
    public int getDistance(Searchable target) {
        Point p = ((Terrain) target).current;
        // Green pixel value.
        return (img.getRGB(p.x, p.y) >> 8) & 0xFF;
    }

    @Override
    public double getHeuristic() {
        if (heur == 0) {
            heur = 14 * (abs(current.x - goal.x) + abs(current.y - goal.y));
        }
        return heur;
    }

    @Override
    public List<Searchable> getMoves() {
        List<Searchable> moves = new ArrayList<>();
        for (Point p : Search.STEPS) {
            int dx = current.x + p.x;
            int dy = current.y + p.y;
            if ((dx >= 0 && dx < img.getWidth() && dy >= 0 && dy < img.getHeight())) {
                moves.add(new Terrain(img, new Point(dx, dy), goal));
            }
        }
        return moves;
    }

    @Override
    public double getCost() {
        return cost;
    }

    @Override
    public void setCost(double cost) {
        this.cost = cost;
    }

    @Override
    public boolean equals(Object obj) {
        Point p = ((Terrain) obj).current;
        return current.equals(p);
    }

    @Override
    public int hashCode() {
        return current.hashCode();
    }

    @Override
    public int compareTo(Searchable o) {
        return Double.compare(getCost(), o.getCost());
    }

    public static void main(String[] args) throws Exception {
        BufferedImage img = ImageIO.read(new File("terrain.png"));
        Searchable start = new Terrain(img, new Point(100, 100), new Point(400, 400));
        Search search = new Search();
        Search.Result res = search.bfs(start);
        System.out.println("bfs1=" + res.iterations);
        res = search.astar(start);
        System.out.println("astar1=" + res.iterations);
    }

}

File: a2/python/terrain.py
#!/usr/bin/env python3
from search import (bfs, astar)
from PIL import Image


class Terrain(object):

    def __init__(self, img, current, goal):
        self.img = img
        self.current = current
        self.goal = goal
        self.pix = img.load()
        self._h = None

    def distance(self, target):
        p = target.current
        # Green pixel value.
        return self.pix[p[0], p[1]][2]

    def done(self):
        return self.current == self.goal

    def heuristic(self):
        self._h = self._h or (14 * (abs(self.current[0] - self.goal[0])
                                 + abs(self.current[1] - self.goal[1])))
        return self._h

    def moves(self):
        p = self.current
        for step in [(p[0]+1, p[1]), (p[0]-1, p[1]),
                     (p[0], p[1]+1), (p[0], p[1]-1)]:
            y, x = step
            if 0 <= y < self.img.size[0] and 0 <= x < img.size[1]:
                yield Terrain(self.img, step, self.goal)

    def __eq__(self, o):
        return self.current == o.current

    def __hash__(self):
        return hash(self.current)


def do_astar(terrain):
    goal, cost, iters = astar(terrain)
    print('astar1=%d' % iters)


def do_bfs(terrain):
    goal, cost, iters = bfs(terrain)
    print('bfs1=%d' % iters)


if __name__ == '__main__':
    img = Image.open('terrain.png')
    start = Terrain(img, (100, 100), (400, 400))
    do_bfs(start)
    do_astar(start)
